### ISSUE TITLE
Wrong diagnostic for rayon trait bound requirements

### ISSUE THREAD

**Issue by harudagondi** [rust-lang/rust#111814]
### Code

```Rust
use rayon::prelude::IntoParallelIterator;

struct A<T> {
    a: Vec<T>,
}

impl<T> A<T> {
    fn foo(self) {
        let x = self.a
            .into_par_iter()
            .map(|_| 1)
            .sum();
        println!("{x}");
    }
}


playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4cc9dae2833d21d973e5656b384dddac
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0599]: the method `into_par_iter` exists for struct `Vec<T>`, but its trait bounds were not satisfied
  --> src/lib.rs:10:14
   |
9  |           let x = self.a
   |  _________________-
10 | |             .into_par_iter()
   | |             -^^^^^^^^^^^^^ method cannot be called on `Vec<T>` due to unsatisfied trait bounds
   | |_____________|
   | 
   |
   = note: the following trait bounds were not satisfied:
           `[T]: Sized`
           which is required by `[T]: IntoParallelIterator`
           `[T]: ParallelIterator`
           which is required by `[T]: IntoParallelIterator`

warning: unused import: `rayon::prelude::IntoParallelIterator`
 --> src/lib.rs:1:5
  |
1 | use rayon::prelude::IntoParallelIterator;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

For more information about this error, try `rustc --explain E0599`.
warning: `playground` (lib) generated 1 warning
error: could not compile `playground` due to previous error; 1 warning emitted
```

### Desired output

```Shell
With code like

trait Trait {
    fn frobnicate(self);
}

impl<T: Send> Trait for Vec<T> {
    fn frobnicate(self) {}
}

struct A<T>(Vec<T>);
    
impl<T> A<T> {
    fn foo(self) { self.0.frobnicate() }
}


It outputs:

```rust
   Compiling playground v0.0.1 (/playground)
error[E0599]: the method `frobnicate` exists for struct `Vec<T>`, but its trait bounds were not satisfied
  --> src/lib.rs:12:27
   |
12 |     fn foo(self) { self.0.frobnicate() }
   |                           ^^^^^^^^^^
  --> /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/alloc/src/vec/mod.rs:400:1
   |
   = note: doesn't satisfy `Vec<T>: Trait`
   |
note: trait bound `T: Send` was not satisfied
  --> src/lib.rs:5:9
   |
5  | impl<T: Send> Trait for Vec<T> {
   |         ^^^^  -----     ------
   |         |
   |         unsatisfied trait bound introduced here

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```
```


### Rationale and extra context

There's nothing in the former error message that indicates that bounding `T: Send` is actually the correct answer.

### Other cases

_No response_

### Anything else?

_No response_
